### PR TITLE
Add upload snapshot dialog

### PR DESCRIPTION
**Pull request by ludeeus**
## Proposed change
 
Adds a new component for uploading snapshots and a new dialog to do so in the snapshot tab of the supervisor.

Related PR's:

- https://github.com/home-assistant/supervisor/pull/2076
- https://github.com/home-assistant/core/pull/40503

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/frontend/issues/3910
- Link to documentation pull request:

snapshot

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  TemplateResult,
+} from "lit-element";
+import { fireEvent } from "../../../../src/common/dom/fire_event";
+import "../../../../src/components/ha-circular-progress";
+import { createCloseHeading } from "../../../../src/components/ha-dialog";
+import "../../../../src/components/ha-formfield";
+import "../../../../src/components/ha-header-bar";
+import "../../../../src/components/ha-radio";
+import "../../../../src/components/ha-related-items";
+import "../../../../src/components/ha-svg-icon";
```
Comment (zsarnett) [inline hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts:21]: not used

File: hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  CSSResult,
+  customElement,
+  html,
+  internalProperty,
+  LitElement,
+  property,
+  TemplateResult,
+} from "lit-element";
+import { fireEvent } from "../../../../src/common/dom/fire_event";
+import "../../../../src/components/ha-circular-progress";
```
Comment (zsarnett) [inline hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts:15]: not used

File: hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts (hunk 1 above)
Comment (ludeeus) [reply hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts]: Thanks :+1:

File: hassio/src/components/hassio-upload-snapshot.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                </label>
+                <iron-input slot="input">
+                  <input
+                    id="input"
+                    type="file"
+                    class="file"
+                    accept="application/x-tar"
+                    @change=${this._handleFilePicked}
+                  />
+                </iron-input>
```
Comment (bramkragten) [inline hassio/src/components/hassio-upload-snapshot.ts:81]: Can we take the guts of the file uploader out of here in a separate element and use it in both picture upload and this uploader?

Comment (bramkragten) [reply hassio/src/components/hassio-upload-snapshot.ts:81]: Like all the handledrop, handle drag and the paper input elements

File: src/data/hassio/snapshot.ts [hunk 4]
```diff
@@ -79,3 +79,21 @@ export const createHassioPartialSnapshot = async (
...
+  hass: HomeAssistant,
+  file: File
+): Promise<HassioSnapshot> => {
+  const fd = new FormData();
+  fd.append("file", file);
+  const resp = await hass.fetchWithAuth("/api/hassio/snapshots/new/upload", {
+    method: "POST",
+    body: fd,
+  });
+  if (resp.status === 413) {
```
Comment (bramkragten) [inline src/data/hassio/snapshot.ts:93]: What is the limit here?

File: hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      <ha-dialog
+        open
+        scrimClickAction
+        escapeKeyAction
+        hideActions
+        @closed=${this.closeDialog}
+        .heading=${createCloseHeading(this.hass, "Upload snapshot")}
+      >
+        <hassio-upload-snapshot
+          @snapshot-uploaded=${this.closeDialog}
```
Comment (bramkragten) [inline hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts:53]: Can we show the detail dialog of the uploaded snapshot after upload is completed?

File: hassio/src/components/hassio-upload-snapshot.ts (hunk 3 above)
Comment (ludeeus) [reply hassio/src/components/hassio-upload-snapshot.ts:81]: Sure :+1:

File: src/data/hassio/snapshot.ts (hunk 4 above)
Comment (ludeeus) [reply src/data/hassio/snapshot.ts:93]: 1024MB

File: hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts (hunk 5 above)
Comment (ludeeus) [reply hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts:53]: Yes, do we want to?

Comment (bramkragten) [reply hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts:53]: I guess the reason for uploading would be to restore it?

Comment (ludeeus) [reply hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts:53]: Seems too logical :smiley:

File: src/data/hassio/snapshot.ts (hunk 4 above)
Comment (bramkragten) [reply src/data/hassio/snapshot.ts:93]: OK, I guess that should be good enough as long as you don't add the media dir 😄

File: hassio/src/components/hassio-upload-snapshot.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    `;
+  }
+
+  private async _uploadFile(ev) {
+    const file = ev.detail.file;
+
+    if (!["application/x-tar"].includes(file.type)) {
+      showAlertDialog(this, {
+        title: "Unsupported file format",
+        text: " Please choose a tar file.",
```
Comment (bramkragten) [inline hassio/src/components/hassio-upload-snapshot.ts:57]:
```suggestion
        text: "Please choose a Home Assistant snapshot file (.tar)",
```

File: src/components/ha-file-upload.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  public render(): TemplateResult {
+    return html`
+      ${this.uploading
+        ? html`<ha-circular-progress
+            alt="Uploading"
+            size="large"
+            active
+          ></ha-circular-progress>`
+        : html`
+            ${this._error ? html`<div class="error">${this._error}</div>` : ""}
```
Comment (bramkragten) [inline src/components/ha-file-upload.ts:62]: This should not be in here I think?

Comment (ludeeus) [reply src/components/ha-file-upload.ts:62]: You are right, error is not used

File: src/components/ha-file-upload.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+            </label>
+          `}
+    `;
+  }
+
+  private _handleDrop(ev: DragEvent) {
+    ev.preventDefault();
+    ev.stopPropagation();
+    if (ev.dataTransfer?.files) {
+      fireEvent(this, "file-added", { file: ev.dataTransfer.files[0] });
```
Comment (bramkragten) [inline src/components/ha-file-upload.ts:105]: Should we prepare this for multi file upload?
```suggestion
      fireEvent(this, "file-picked", { files: ev.dataTransfer.files });
```

Comment (ludeeus) [reply src/components/ha-file-upload.ts:105]: Sure :+1:

**Review by bramkragten — APPROVED** [review]

File: hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts (hunk 5 above)
Comment (ludeeus) [reply hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts]: This also makes a lot of sense if added elsewhere (like during onboarding) :+1:

**Review by KTibow — APPROVED** [review]

**Pull request merged**